File: 1447-jump-game-iv/jump-game-iv.cpp
class Solution {
public:
    int minJumps(vector<int>& arr) {
        int n = arr.size();

        // Edge case
        if (n == 1)
            return 0;

        // Store all indices for each value
        unordered_map<int, vector<int>> mp;

        for (int i = 0; i < n; i++) {
            mp[arr[i]].push_back(i);
        }

        // BFS
        queue<int> q;
        vector<bool> visited(n, false);

        q.push(0);
        visited[0] = true;

        int steps = 0;

        while (!q.empty()) {

            int size = q.size();

            while (size--) {

                int curr = q.front();
                q.pop();

                // Reached last index
                if (curr == n - 1)
                    return steps;

                // Store neighbors
                vector<int> nextIndices;

                // Same value jumps
                for (int idx : mp[arr[curr]]) {
                    nextIndices.push_back(idx);
                }

                // i + 1
                if (curr + 1 < n)
                    nextIndices.push_back(curr + 1);

                // i - 1
                if (curr - 1 >= 0)
                    nextIndices.push_back(curr - 1);

                // Visit neighbors
                for (int next : nextIndices) {

                    if (!visited[next]) {
                        visited[next] = true;
                        q.push(next);
                    }
                }

                // Important optimization
                mp[arr[curr]].clear();
            }

            steps++;
        }

        return -1;
    }
};

File: 1447-jump-game-iv/jump-game-iv.py
import collections
from typing import List

class Solution:
    def minJumps(self, arr: List[int]) -> int:
        n = len(arr)
        # Base case: if array has only 1 element, we're already at the end.
        if n <= 1:
            return 0
        
        # Dictionary to store value -> list of indices
        graph = collections.defaultdict(list)
        for i, val in enumerate(arr):
            graph[val].append(i)
            
        # Initialize BFS
        queue = collections.deque([0])  # store index
        visited = {0}
        steps = 0
        
        while queue:
            # Level-by-level traversal
            for _ in range(len(queue)):
                curr = queue.popleft()
                
                # If we've reached the last index, return the step count
                if curr == n - 1:
                    return steps
                
                # Gather all possible next indices: j (same value), i-1, i+1
                neighbors = graph[arr[curr]]
                neighbors.append(curr - 1)
                neighbors.append(curr + 1)
                
                # Check bounds and if visited
                for nxt in neighbors:
                    if 0 <= nxt < n and nxt not in visited:
                        visited.add(nxt)
                        queue.append(nxt)
                
                # CRITICAL STEP: Clear the teleport list for this value. 
                # This prevents O(N^2) worst-case scenarios and avoids TLE.
                graph[arr[curr]].clear()
                
            steps += 1
            
        return -1